plutus-core/plutus-ir/src/PlutusIR/Compiler.hs

```diff
@@ -179,22 +216,44 @@ simplifyTerm = runIfOpts simplify'
 -- nearest lambda/Lambda/letStrictNonValue.
 -- Note: It assumes globally unique names
 floatOut
-    :: (Compiling m e uni fun a, Semigroup b)
+    :: (Compiling m e uni fun a, Semigroup b, b ~ Provenance a)
     => Term TyName Name uni fun b
     -> m (Term TyName Name uni fun b)
 floatOut t = do
     binfo <- view ccBuiltinsInfo
-    runIfOpts (pure . LetMerge.letMerge . RecSplit.recSplit . LetFloatOut.floatTerm binfo) t
+    flip runIfOpts t $ do
+      pure . LetFloatOut.floatTerm binfo
+      >=> updateCertTrace PassFloatOut
+      >=> pure . RecSplit.recSplit
+      >=> updateCertTrace PassRecSplit
+      >=> pure . LetMerge.letMerge
+      >=> updateCertTrace PassLetMerge
 
 -- | Perform floating in/merging of lets in a 'Term'.
 floatIn
-    :: Compiling m e uni fun a
+    :: (Compiling m e uni fun a, b ~ Provenance a)
     => Term TyName Name uni fun b
     -> m (Term TyName Name uni fun b)
 floatIn t = do
     binfo <- view ccBuiltinsInfo
     relaxed <- view (ccOpts . coRelaxedFloatin)
-    runIfOpts (fmap LetMerge.letMerge . LetFloatIn.floatTerm binfo relaxed) t
+    flip runIfOpts t $ do
+      PLC.rename
+      >=> updateCertTrace PassRename
```

Comment:
This will be easier to do with the Pass abstraction: Float-in would be a `CompoundPass` that consists of two Passes: renaming and float-in proper. This drive module does not need to know such details.